File: java/Samples/Biometrics/VB/AbisSampleVB/Common/BusyIndicator.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Public Class BusyIndicator
	Inherits Control
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		DoubleBuffered = True
		timer.Start()
	End Sub

#End Region

#Region "Private fields"

	Private WithEvents timer As Timer
	Private components As IContainer
	Private currentAngle As Integer = 0

#End Region

#Region "Private constants"

	Private Const Radius As Integer = 100
	Private Const CircleRadius As Integer = 33
	Private Const BorderMargin As Integer = 4
	Private Const Center As Integer = BorderMargin + CircleRadius + Radius
	Private Const CombindedSize As Single = BorderMargin + CircleRadius + Radius * 2 + CircleRadius + BorderMargin

#End Region

#Region "Private methods"

	Private Function GetRectangle(ByVal angle As Double) As RectangleF
		Dim x As Double = Center + Radius * Math.Cos(angle)
		Dim y As Double = Center + Radius * Math.Sin(angle)

		Return New RectangleF(CSng(x) - CircleRadius, CSng(y) - CircleRadius, CircleRadius * 2, CircleRadius * 2)
	End Function

	Private Sub InitializeComponent()
		Me.components = New System.ComponentModel.Container()
		Me.timer = New System.Windows.Forms.Timer(Me.components)
		Me.SuspendLayout()
		' 
		' timer
		' 
		Me.timer.Interval = 33
		Me.ResumeLayout(False)
	End Sub

	Private Sub TimerTick(ByVal sender As Object, ByVal e As EventArgs) Handles timer.Tick
		currentAngle = (currentAngle + 10) Mod 360
		Invalidate()
	End Sub

#End Region

#Region "Protected methods"

	Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
		MyBase.OnPaint(e)

		Dim g As Graphics = e.Graphics
		g.SmoothingMode = SmoothingMode.AntiAlias

		Dim sz As Size = Me.Size
		Dim zoom As Single = Math.Min(sz.Width / CombindedSize, sz.Height / CombindedSize)
		Dim dx As Single = (zoom * CombindedSize - sz.Width) / 2
		Dim dy As Single = (zoom * CombindedSize - sz.Height) / 2

		Dim m As Matrix = g.Transform
		m.RotateAt(currentAngle, New PointF(sz.Width \ 2, sz.Height \ 2))
		g.Transform = m

		g.TranslateTransform(-dx, -dy)
		g.ScaleTransform(zoom, zoom)

		Dim value As Byte = 0
		For angle As Double = 0 To 2 * Math.PI - 1 Step Math.PI / 4
			value += 25
			Using b = New SolidBrush(Color.FromArgb(value, value, value))
				g.FillEllipse(b, GetRectangle(angle))
			End Using
		Next angle
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/Common/BusyIndicator.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace Neurotec.Samples
{
	public class BusyIndicator : Control
	{
		#region Public constructor

		public BusyIndicator()
		{
			InitializeComponent();
			DoubleBuffered = true;
			timer.Start();
		}

		#endregion

		#region Private fields

		private Timer timer;
		private IContainer components;
		private int currentAngle = 0;

		#endregion

		#region Private constants

		private const int Radius = 100;
		private const int CircleRadius = 33;
		private const int BorderMargin = 4;
		private const int Center = BorderMargin + CircleRadius + Radius;
		private const float CombindedSize = BorderMargin + CircleRadius + Radius * 2 + CircleRadius + BorderMargin;

		#endregion

		#region Private methods

		private RectangleF GetRectangle(double angle)
		{
			double x = Center + Radius * Math.Cos(angle);
			double y = Center + Radius * Math.Sin(angle);

			return new RectangleF((float)x - CircleRadius, (float)y - CircleRadius, CircleRadius * 2, CircleRadius * 2);
		}

		private void InitializeComponent()
		{
			this.components = new System.ComponentModel.Container();
			this.timer = new System.Windows.Forms.Timer(this.components);
			this.SuspendLayout();
			// 
			// timer
			// 
			this.timer.Interval = 33;
			this.timer.Tick += new System.EventHandler(this.TimerTick);
			this.ResumeLayout(false);
		}

		private void TimerTick(object sender, EventArgs e)
		{
			currentAngle = (currentAngle + 10) % 360;
			Invalidate();
		}

		#endregion

		#region Protected methods

		protected override void OnPaint(PaintEventArgs e)
		{
			base.OnPaint(e);

			Graphics g = e.Graphics;
			g.SmoothingMode = SmoothingMode.AntiAlias;

			Size sz = this.Size;
			float zoom = Math.Min(sz.Width / CombindedSize, sz.Height / CombindedSize);
			float dx = (zoom * CombindedSize - sz.Width) / 2;
			float dy = (zoom * CombindedSize - sz.Height) / 2;

			Matrix m = g.Transform;
			m.RotateAt(currentAngle, new PointF(sz.Width / 2, sz.Height / 2));
			g.Transform = m;

			g.TranslateTransform(-dx, -dy);
			g.ScaleTransform(zoom, zoom);

			byte value = 0;
			for (double angle = 0; angle < 2 * Math.PI; angle += Math.PI / 4)
			{
				value += 25;
				using (var b = new SolidBrush(Color.FromArgb(value, value, value)))
					g.FillEllipse(b, GetRectangle(angle));
			}
		}

		#endregion
	}
}
